snuba/query/snql/parser.py

```diff
@@ -632,11 +646,65 @@ def parse(exp: Expression) -> Expression:
     query.transform_expressions(parse)
 
 
+ARRAY_JOIN_MATCH = FunctionCallMatch(
+    StringMatch("arrayJoinExplode"),
+    (
+        Param("column", ColumnMatch(AnyOptionalString(), AnyMatch(str))),
+        Param("op", Or([LiteralMatch(StringMatch(op)) for op in OPERATOR_TO_FUNCTION])),
+        Param("value", AnyExpression()),
+    ),
+)
+
+
+def _array_join_transformation(query: LogicalQuery) -> None:
+    def parse(exp: Expression) -> Expression:
+        result = ARRAY_JOIN_MATCH.match(exp)
+        if result:
+            column = result.expression("column")
+            assert isinstance(column, Column)
+
+            op_literal = result.expression("op")
+            assert isinstance(op_literal, Literal)
+            op = str(op_literal.value)
+
+            value = result.expression("value")
+
+            function_name = "arrayExists" if op in POSITIVE_OPERATORS else "arrayAll"
+
+            return FunctionCall(
+                None,
+                function_name,
+                (
+                    Lambda(
+                        None,
+                        ("x",),
+                        FunctionCall(
+                            None,
+                            "assumeNotNull",
+                            (
+                                FunctionCall(
+                                    None,
+                                    OPERATOR_TO_FUNCTION[op],
+                                    (Argument(None, "x"), value,),
+                                ),
+                            ),
+                        ),
+                    ),
+                    column,
+                ),
+            )
+
+        return exp
+
+    query.transform_expressions(parse)
+
+
 def parse_snql_query(body: str, dataset: Dataset) -> LogicalQuery:
     query = parse_snql_query_initial(body)
 
     # These are the post processing phases
     _parse_datetime_literals(query)
+    _array_join_transformation(query)
```

Comment:
Since this is really equivalent to a custom function, could we move it to the end ?

Reply:
I don't think this was addressed. Is there any reason not to move this, which is logically a custom function, at the end of the list of these processors, when the query has been fully sanitized ?